src/container/context-stack.ts

```diff
@@ -0,0 +1,66 @@
+import { ConfigurableBindingScopeEnum, ContextHierarchyOptionEnum } from "../constants/literal_types";
+import { interfaces } from "../interfaces/interfaces";
+import { getBindingDictionary } from "../planning/planner";
+import { Stack } from "./stack";
+
+export class ContextStack {
+  constructor(private readonly container:interfaces.Container){}
+  private _contextStack:interfaces.Stack<interfaces.Context> = new Stack<interfaces.Context>();
+  private _contextStackDepth = 0;
+  private _requiresContextStack = false;
+  planCompleted(context:interfaces.Context):void {
+    this._contextStackDepth++;
+    if(this._contextStackDepth === 1) {
+        this._determineRequiresContextHierarchy();
+    }
+    if (this._requiresContextStack) {
+        context.parentContext = this._contextStack.peek();
+        this._contextStack.push(context);
+    }
+  }
+  resolved():void {
+    this._contextStackDepth--;
+    this._contextStack.pop();
+  }
+
+  inRootRequestScope(context:interfaces.Context) {
+    this._contextStack.push(context);
+  }
+
+  private _determineRequiresContextHierarchy(): void {
+    let requiresContextHierarchyOption: interfaces.ContextHierarchyOption = ContextHierarchyOptionEnum.Allow;
```

Comment:
Defaults to Allow.  I wanted it to be IfBindedInCustomOrRootRequestScope but the performance test 
`it("Should be able to register 10K bindings in less than 1 ms", () => `
failed as went to 3s.

Thoughts on default ?